src/cmd/ksh93/sh/lex.c

```diff
@@ -57,8 +57,7 @@
 
 #define oldmode(lp) (lp->lexd.lastc >> CHAR_BIT)
 #define endchar(lp) (lp->lexd.lastc & 0xff)
-#define setchar(lp, c) (lp->lexd.lastc = ((lp->lexd.lastc & ~0xff) | (c)))
-#define poplevel(lp) (lp->lexd.lastc = lp->lexd.lex_match[--lp->lexd.level])
+#define setchar(lp, c) lp->lexd.lastc = ((lp->lexd.lastc & ~0xff) | (c))
```

Comment:
Not sure why the brackets have been removed from this macro (setchar).  I thought that it was a good idea to surround macros with brackets to prevent side-effects?

Reply:
Yes, it was.

Reply:
@dannyweldon Broadly speaking there are two types of macros: a) those that return a rvalue and can be used in an expression and b) those meant to be used like a void returning function call. The `setchar()`, `poplevel()` and `pushlevel()` are type b macros. It is a mistake to use them in an expression such as `if (x && setchar())`. The parens around a type b macro are irrelevant. In fact, traditionally such macros are written using this pattern to ensure that if they are used in an expression it will fail to compile:

```
#define setchar(lp, c) do { something; } while (0)
```

Today I would write that as a `static inline void` function. I already changed `poplevel()` and `pushlevel()` to functions and intended to do the same to `setchar()` but got sidetracked. Thanks for the reminder to finish that cleanup.

P.S., Yes, macros are used in other ways as well. For example, it's less common but sometimes a macro is meant to be used as an lvalue. There are other highly stylized patterns where a macro can't be substituted with a static inline function but you also don't want it to be usable in an expression. See, for example, `SFMTXRETURN`.

Reply:
Wow.  Thanks for the detailed explanation, @krader1961.

Reply:
> The parens around a type b macro are irrelevant.

They are not irrelevant in general.  When such macros are used in code, they look like function calls.  If you remove the brackets, such seeming function calls may have completely unexpected side effects in certain cases.  By removing the protection that was already there, you can shoot yourself in the foot later on.